File: backend/src/services/rag/vectorStore.ts
import { getPineconeIndex }           from "../../config/pinecone";
import { TextChunk, RetrievedChunk }  from "../../types";
import { logger }                     from "../../config/logger";
import { env }                        from "../../config/env";

const BATCH = 100;

const isPineconeReady = (): boolean => {
  try {
    getPineconeIndex();
    return true;
  } catch {
    return false;
  }
};

export const upsertChunks = async (chunks: TextChunk[]): Promise<number> => {
  // Mock mode when Pinecone is not available or initialization failed
  if (!env.PINECONE_API_KEY || !isPineconeReady()) {
    logger.info(`[MOCK] Mock upserting ${chunks.length} chunks...`);
    return chunks.length;
  }

  const index   = getPineconeIndex();
  const vectors = chunks
    .filter((c) => c.vector)
    .map((c) => ({
      id:       c.chunkId,
      values:   c.vector!,
      metadata: c.metadata,
    }));

  let total = 0;
  for (let i = 0; i < vectors.length; i += BATCH) {
    const batch = vectors.slice(i, i + BATCH);
    await index.upsert(batch);
    total += batch.length;
    logger.info(`Upserted batch: ${batch.length} vectors`);
  }
  return total;
};

export const similaritySearch = async (
  queryVector: number[],
  documentId:  string,
  userId:      string,
  topK = 5
): Promise<RetrievedChunk[]> => {
  // Mock mode when Pinecone is not available or initialization failed
  if (!env.PINECONE_API_KEY || !isPineconeReady()) {
    logger.info(`[MOCK] Mock similarity search returning ${topK} results...`);
    return Array.from({ length: topK }, (_, i) => ({
      chunkId: `mock-chunk-${i}`,
      score: 0.8 - (i * 0.1),
      text: `Mock chunk ${i + 1}: This is sample text from the document that would be relevant to the query.`,
    }));
  }

  const index   = getPineconeIndex();
  const results = await index.query({
    vector:          queryVector,
    topK,
    includeMetadata: true,
    filter: {
      documentId: { $eq: documentId },
      userId:     { $eq: userId },
    },
  });

  return (results.matches ?? []).map((m) => ({
    chunkId: m.id,
    score:   Math.round((m.score ?? 0) * 10000) / 10000,
    text:    (m.metadata?.["text"] as string) ?? "",
  }));
};

export const deleteDocumentVectors = async (
  documentId: string
): Promise<void> => {
  // Mock mode when Pinecone is not available or initialization failed
  if (!env.PINECONE_API_KEY || !isPineconeReady()) {
    logger.info(`[MOCK] Mock deleting vectors for doc: ${documentId}`);
    return;
  }

  const index = getPineconeIndex();
  await index.deleteMany({ documentId: { $eq: documentId } });
  logger.info(`Deleted vectors for doc: ${documentId}`);
};

File: backend/app/services/vector_store.py
import logging
from typing import List

from app.core.pinecone_client import get_pinecone_index

logger = logging.getLogger(__name__)

BATCH_SIZE = 100


def upsert_chunks(chunks: List[dict]) -> int:
    index = get_pinecone_index()
    if index is None:
        logger.warning("[vector_store] Pinecone not available, skipping upsert")
        return 0

    vectors = [
        {"id": c["chunk_id"], "values": c["vector"], "metadata": c["metadata"]}
        for c in chunks
        if c.get("vector")
    ]

    total = 0
    for i in range(0, len(vectors), BATCH_SIZE):
        batch = vectors[i : i + BATCH_SIZE]
        index.upsert(vectors=batch)
        total += len(batch)
        logger.info("[vector_store] Upserted %d vectors", len(batch))

    return total


def similarity_search(
    query_vector: List[float],
    document_id: str,
    user_id: str,
    top_k: int = 5,
) -> List[dict]:
    index = get_pinecone_index()
    if index is None:
        logger.warning("[vector_store] Pinecone not available")
        return []

    results = index.query(
        vector=query_vector,
        top_k=top_k,
        include_metadata=True,
        filter={
            "documentId": {"$eq": document_id},
            "userId": {"$eq": user_id},
        },
    )

    return [
        {
            "chunk_id": match["id"],
            "score": round(match.get("score", 0), 4),
            "text": match.get("metadata", {}).get("text", ""),
        }
        for match in results.get("matches", [])
    ]


def delete_document_vectors(document_id: str) -> None:
    index = get_pinecone_index()
    if index is None:
        return
    try:
        index.delete(filter={"documentId": {"$eq": document_id}})
        logger.info("[vector_store] Deleted vectors for doc: %s", document_id)
    except Exception as exc:
        logger.error("[vector_store] Delete error: %s", exc)
